--- a/Makefile.am	2019-06-08 23:09:44.259895000 +0300
+++ b/Makefile.am.new	2019-06-09 01:36:22.236256879 +0300
@@ -979,7 +979,7 @@
 dist_noinst_SCRIPTS += test/run_vlc.sh
 
 ###############################################################################
-# Installing plugins cache
+# Installing plugins cache and fix acestreamplayer build names
 ###############################################################################
 install-exec-hook:
 	if test "$(build)" = "$(host)"; then \
@@ -989,6 +989,45 @@
 			 "$(DESTDIR)$(vlclibdir)/plugins" ; \
 	else \
 		echo "Cross-compilation: cache generation skipped!" ; \
+	fi && \
+	echo "Fix Acestream files"  && \
+	if [ -f /usr/bin/vlc ]; then \
+		cp -f /usr/bin/vlc /usr/bin/acestreamplayer; \
+	fi && \
+	if [ -f /usr/bin/vlc-wrapper ]; then \
+		cp -f /usr/bin/vlc-wrapper /usr/bin/acestreamplayer-wrapper; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/libvlc.pc ]; then \
+		cp -f /usr/lib64/pkgconfig/libvlc.pc /usr/lib64/pkgconfig/libacestreamplayer.pc; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/vlc-plugin.pc ]; then \
+		cp -f /usr/lib64/pkgconfig/vlc-plugin.pc /usr/lib64/pkgconfig/acestreamplayer-plugin.pc; \
+	fi && \
+	echo "Restore VLC files"  && \
+	if [ -f /usr/bin/_vlc ]; then \
+		cp -f /usr/bin/_vlc /usr/bin/vlc; \
+	fi && \
+	if [ -f /usr/bin/_vlc-wrapper ]; then \
+		cp -f /usr/bin/_vlc-wrapper /usr/bin/vlc-wrapper; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/_libvlc.pc ]; then \
+		cp -f /usr/lib64/pkgconfig/_libvlc.pc /usr/lib64/pkgconfig/libvlc.pc; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/_vlc-plugin.pc ]; then \
+		cp -f /usr/lib64/pkgconfig/_vlc-plugin.pc /usr/lib64/pkgconfig/vlc-plugin.pc; \
+	fi && \
+	echo "Remove backup VLC files"  && \
+	if [ -f /usr/bin/_vlc ]; then \
+		rm -f /usr/bin/_vlc; \
+	fi && \
+	if [ -f /usr/bin/_vlc-wrapper ]; then \
+		rm -f /usr/bin/_vlc-wrapper; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/_libvlc.pc ]; then \
+		rm -f /usr/lib64/pkgconfig/_libvlc.pc; \
+	fi && \
+	if [ -f /usr/lib64/pkgconfig/_vlc-plugin.pc ]; then \
+		rm -f /usr/lib64/pkgconfig/_vlc-plugin.pc; \
 	fi
 
 uninstall-hook:
