--- vlc-2.1.2.orig/modules/gui/qt4/actions_manager.cpp	2013-12-03 11:12:34.000000000 +0200
+++ vlc-2.1.2/modules/gui/qt4/actions_manager.cpp	2014-01-17 16:55:14.874992774 +0200
@@ -35,7 +35,10 @@
 #include "components/controller.hpp" /* Toggle FSC controller width */
 #include "components/extended_panels.hpp"
 
+#include <QDesktopServices>
+#include <QUrl>
+
 ActionsManager::ActionsManager( intf_thread_t * _p_i )
 {
     p_intf = _p_i;
 }
@@ -97,6 +100,41 @@
             if( p_intf->p_sys->p_mi )
                 p_intf->p_sys->p_mi->getFullscreenControllerWidget()->toggleFullwidth();
             break;
+        case SAVE_P2P_ACTION : {
+            input_thread_t *p_input = THEMIM->getInput();
+            if( p_input ) {
+                input_item_t *p_item = input_GetItem( p_input );
+                if( p_item )
+                    THEDP->p2p_SaveDialog( p_item );
+            }
+            else {
+                playlist_item_t *p_last_playing_item = playlist_LastPlayingItem(THEPL);
+                if( p_last_playing_item && p_last_playing_item->p_input ) {
+                    THEDP->p2p_SaveDialog( p_last_playing_item->p_input );
+                }
+            }
+            break;
+        }
+        case VISIT_AD_ACTION: {
+            QString url = THEMIM->getIM()->getVisitUrl();
+            if(!url.isEmpty()) {
+                bool fs = var_GetBool( THEPL, "fullscreen" );
+                if(fs)
+                    fullscreen();
+                QDesktopServices::openUrl(QUrl(url));
+            }
+            break;
+        }
+        case DISABLE_AD_ACTION: {
+            QString url = THEMIM->getIM()->getDisableUrl();
+            if(!url.isEmpty()) {
+                bool fs = var_GetBool( THEPL, "fullscreen" );
+                if(fs)
+                    fullscreen();
+                QDesktopServices::openUrl(QUrl(url));
+            }
+            break;
+        }
         default:
             msg_Dbg( p_intf, "Action: %i", id_action );
             break;
