--- vlc-2.1.2.orig/modules/gui/qt4/dialogs/mediainfo.cpp	2013-06-24 21:00:38.000000000 +0300
+++ vlc-2.1.2/modules/gui/qt4/dialogs/mediainfo.cpp	2014-01-17 15:54:28.994989652 +0200
@@ -63,6 +63,8 @@
     infoTabW->insertTab( META_PANEL, MP, qtr( "&General" ) );
     EMP = new ExtraMetaPanel( infoTabW );
     infoTabW->insertTab( EXTRAMETA_PANEL, EMP, qtr( "&Metadata" ) );
+    P2PMP = new P2PPanel( infoTabW, p_intf );
+    infoTabW->insertTab( P2P_PANEL, P2PMP, qtr( P2P_PRODUCT_NAME ) );
     IP = new InfoPanel( infoTabW );
     infoTabW->insertTab( INFO_PANEL, IP, qtr( "Co&dec" ) );
     if( isMainInputInfo )
@@ -116,6 +118,8 @@
         DCONNECT( THEMIM->getIM(), currentMetaChanged( input_item_t* ),
                   EMP, update( input_item_t* ) );
         DCONNECT( THEMIM->getIM(), statisticsUpdated( input_item_t* ),
+                  P2PMP, update( input_item_t* ) );
+        DCONNECT( THEMIM->getIM(), statisticsUpdated( input_item_t* ),
                   ISP, update( input_item_t* ) );
 
         if( THEMIM->getInput() )
@@ -153,6 +157,7 @@
     IP->update( p_item );
     MP->update( p_item );
     EMP->update( p_item );
+    P2PMP->update( p_item );
 
     if( isMainInputInfo ) ISP->update( p_item );
 }
@@ -162,6 +167,7 @@
     IP->clear();
     MP->clear();
     EMP->clear();
+    P2PMP->clear();
 
     if( isMainInputInfo ) ISP->clear();
 }
