--- a/extras/package/win32/package.mak	2015-01-22 16:27:37.000000000 +0300
+++ b/extras/package/win32/package.mak.new	2019-06-20 01:34:59.974209100 +0300
@@ -11,9 +11,9 @@
 
 
 if HAVE_WIN32
-include extras/package/npapi.am
+#include extras/package/npapi.am
 
-build-npapi: package-win-install
+#build-npapi: package-win-install
 endif
 
 if HAVE_WIN64
@@ -75,13 +75,7 @@
 # Remove cruft
 	find $(win32_destdir)/plugins/ -type f \( -name '*.a' -or -name '*.la' \) -exec rm -rvf {} \;
 
-package-win-npapi: build-npapi
-	cp "$(top_builddir)/npapi-vlc/installed/lib/axvlc.dll" "$(win32_destdir)/"
-	cp "$(top_builddir)/npapi-vlc/installed/lib/npvlc.dll" "$(win32_destdir)/"
-	mkdir -p "$(win32_destdir)/sdk/activex/"
-	cd $(top_builddir)/npapi-vlc && cp activex/README.TXT share/test/test.html $(win32_destdir)/sdk/activex/
-
-package-win-strip: package-win-common package-win-npapi
+package-win-strip: package-win-common
 	mkdir -p "$(win32_debugdir)"/
 	cd $(win32_destdir); find . -type f \( -name '*$(LIBEXT)' -or -name '*$(EXEEXT)' \) | while read i; \
 	do if test -n "$$i" ; then \
@@ -110,12 +104,10 @@
 
 
 package-win32-xpi: package-win32-webplugin-common
-	cp $(top_builddir)/npapi-vlc/npapi/package/install.rdf "$(win32_xpi_destdir)/"
 	cd $(win32_xpi_destdir) && zip -r -9 "../$(WINVERSION).xpi" install.rdf plugins
 
 
 package-win32-crx: package-win32-webplugin-common
-	cp $(top_builddir)/npapi-vlc/npapi/package/manifest.json "$(win32_xpi_destdir)/"
 	crxmake --pack-extension "$(win32_xpi_destdir)" \
 		--extension-output "$(win32_destdir)/$(WINVERSION).crx" --ignore-file install.rdf
 
