diff -aurN a/contrib/src/sqlite/rules.mak b/contrib/src/sqlite/rules.mak
--- a/contrib/src/sqlite/rules.mak	2019-06-26 18:18:56.320397000 +0300
+++ b/contrib/src/sqlite/rules.mak	2019-06-11 19:50:12.313587000 +0300
@@ -3,7 +3,7 @@
 SQLITE_VERSION := 3.6.20
 SQLITE_URL := http://www.sqlite.org/sqlite-amalgamation-$(SQLITE_VERSION).tar.gz
 
-# PKGS += sqlite
+PKGS += sqlite
 
 ifeq ($(call need_pkg,"sqlite3"),)
 PKGS_FOUND += sqlite
