--- a/win32/src/Makefile	2019-07-23 19:11:49.693821585 +0300
+++ b/win32/src/Makefile.new	2019-07-23 23:09:18.615571677 +0300
@@ -3209,7 +3209,6 @@
 .rc.in.rc: $(top_builddir)/config.status
 	cd "$(top_builddir)" && \
 	$(SHELL) ./config.status --file="src/$@"
-//
 
 libvlc_win32_rc.$(OBJEXT): libvlc_win32_rc.rc $(top_srcdir)/extras/package/win32/libvlc.dll.manifest
 	$(WINDRES) --include-dir $(top_srcdir)/share --include-dir $(top_srcdir)/extras/package/win32 -i $< -o $@
